ts/image-occlusion/mask-editor.ts

```diff
@@ -0,0 +1,136 @@
+    const metadata = await getImageClozeMetadata(path!);
+    const b64encoded = protobuf.util.base64.encode(
+        metadata.data,
+        0,
+        metadata.data.length,
+    );
+    const data = "data:image/png;base64," + b64encoded;
+
+    const canvas = new fabric.Canvas("canvas", {
+        hoverCursor: "pointer",
+        selectionBorderColor: "green",
+    });
+
+    // get image width and height
+    const image = new Image();
+    image.onload = function() {
+        const size = limitSize({ width: image.width, height: image.height });
+        canvas.setWidth(size.width);
+        canvas.setHeight(size.height);
+
+        fabric.Image.fromURL(image.src, function(image) {
+            canvas.setBackgroundImage(image, canvas.renderAll.bind(canvas), {
+                scaleX: canvas.width! / image.width!,
+                scaleY: canvas.height! / image.height!,
+            });
+
+            canvas.backgroundImage.scale(size.scalar);
+            const zoomRatio = (innerWidth - 40) / canvas.width!;
+            zoomResetValue.set(zoomRatio);
+            instance.smoothZoom(0, 0, zoomRatio);
+        });
+    };
+
+    tagsWritable.set([]);
+    globalThis.canvas = canvas;
+    undoRedoInit(canvas);
+    image.src = data;
+    image.remove();
+    return canvas;
+};
+
+export const setupMaskEditorForEdit = async (noteId: number, instance: PanZoom): Promise<fabric.Canvas> => {
+    const clozeNote: ImageOcclusion.ImageClozeNote = await getImageClozeNotes(noteId);
+    const b64encoded = protobuf.util.base64.encode(
+        clozeNote.imageData,
+        0,
+        clozeNote.imageData.length,
+    );
+    const data = "data:image/png;base64," + b64encoded;
+
+    const canvas = new fabric.Canvas("canvas", {
+        hoverCursor: "pointer",
+        selectionBorderColor: "green",
+    });
+
+    // get image width and height
+    const image = new Image();
+    image.onload = function() {
+        const size = limitSize({ width: image.width, height: image.height });
+        canvas.setWidth(size.width);
+        canvas.setHeight(size.height);
+
+        fabric.Image.fromURL(image.src, function(image) {
+            canvas.setBackgroundImage(image, canvas.renderAll.bind(canvas), {
+                scaleX: canvas.width / image.width,
+                scaleY: canvas.height / image.height,
+            });
+
+            canvas.backgroundImage.scale(size.scalar);
+            generateShapeFromCloze(canvas, clozeNote.occlusions);
+            enableSelectable(canvas, true);
+            addClozeNotesToTextEditor(clozeNote.header, clozeNote.backExtra, clozeNote.tags);
+
+            const zoomRatio = (innerWidth - 40) / canvas.width!;
+            zoomResetValue.set(zoomRatio);
+            instance.smoothZoom(0, 0, zoomRatio);
+        });
+    };
+
+    globalThis.canvas = canvas;
+    undoRedoInit(canvas);
+    image.src = data;
+    image.remove();
+    return canvas;
```

Comment:
There is a lot of duplicate code here; can the common parts be factored out into a separate function?